Revision based on feedback (#3)
mihaibudiu/sigmod-rh: streams.tex

\section{Stream operators}\label{sec:streams}

%\subsection{Useful stream operators}\label{sec:abelian}

For the rest of this paper we require the set of values $A$ of a
stream $\stream{A}$ to form a commutative group, with operations $+$,
$-$, and a $0$ (zero) value.  The \emph{plus} defines what it means to
\emph{add} new data, while the \emph{minus} allows us to compute
differences (deltas).  We show later that this requirement is not a
problem for using \dbsp in the context of databases.

Stream operators are very powerful mathematically, but in \dbsp we
restrict ourselves to a very small subset.  All \dbsp computations are
\emph{causal}~\cite{causal}: the output at time $t$ is produced

> immediately after all inputs at time $t$ have been received.  This

Reviewer: Should this say more explicitly that the output at time t depends only on inputs at times <= t?

\end{tikzpicture}
\end{center}
%\end{definition}
%
While this definition may seem strange, because the output stream is
used to compute itself, the use of the delay in the ``feedback'' loop
ensures that only ``previous'' values of the output are used in
computing the current one.  Using the notation $o = \I(s)$ to make
formulas more readable, we can see the contents of stream $o$ is
produced step by step:
\begin{align*}
  o[0] &= s[0] + (\zm(o))[0] = s[0] + 0 = s[0] \\
  o[1] &= s[1] + (\zm(o))[1] = s[1] + o[0] = s[1] + s[0] \\

>   o[2] &= s[2] + (\zm(o))[2] = s[2] + o[1] = s[2] + (s[1] + s[0])

Reviewer: ```suggestion
  o[2] &= s[2] + (\zm(o))[2] = s[2] + o[1] = s[2] + s[1] + s[0]
```

%$\forall s . \I(\D(s)) = \D(\I(s)) = s$.
%\end{theorem}

Integration and differentiation are inverses of each other: while $\D$

> compute the changes of a stream, $\I$ reconstitutes the original

Reviewer: ```suggestion
computes the changes of a stream, $\I$ reconstitutes the original
```

\section{Incremental view maintenance}\label{sec:incremental}

The results in this section are not specific to databases, they are
true for arbitrary streaming computations.  However, for each result
we emphasize how it applies in the domain of databases.  We tie these

> results to databases more tightly in the next section.

Reviewer: ```suggestion
we emphasize how it applies in the domain of databases.  We directly tie these
results to databases in the next section.
```

\end{tikzpicture}
\end{tabular}

\noindent
(We have omitted the labels on the inputs of $T$ for brevity.) In
other words, the incremental version of a feedback loop around a query

> is just the feedback loop with the incremental query for its body.

Reviewer: we haven't defined T^{\Delta} for queries with two inputs.

%that stream operators $+$, $-$, and $\zm$ are identical to their incremental versions.
%The following theorems generalize this to linear and bi-linear operators:

We call an operator $S$ \defined{linear} if it has the property that
$S(a+b) = S(a) + S(b)$ (where $+$ is the addition of streams).

%\begin{theorem}[Linear]\label{linear}
For a linear operator $S$ we have $\inc{S}=S$.
%\end{theorem}

This is very useful because many primitive database operations can be
implemented as linear operators.  We will see that in databases
selection, projection, filtering, grouping, parts of aggregation are
all linear.  Moreover, many of the operators we have seen so far are
linear: $-$, $z^{-1}$, $\I$, $\D$, $\lift{f}$ if $f$ is a linear
function.

> We call an operator with two inputs $T$ \defined{bilinear} if it

Reviewer: ```suggestion
We call an operator $T$ with two inputs \defined{bilinear} if it
```
